platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,231 @@
+import com.swirlds.platform.roster.RosterValidator;
+import com.swirlds.platform.state.RosterStateAccessor;
+import com.swirlds.platform.state.RosterStateModifier;
+import com.swirlds.platform.state.signed.ReservedSignedState;
+import com.swirlds.platform.system.SoftwareVersion;
+import com.swirlds.state.spi.CommittableWritableStates;
+import com.swirlds.state.spi.WritableKVState;
+import com.swirlds.state.spi.WritableSingletonState;
+import com.swirlds.state.spi.WritableStates;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Read-write interface for accessing rosters states.
+ */
+public class WritableRosterStore implements RosterStateModifier {
+
+    private final WritableStates writableStates;
+    private final RosterStateAccessor rosterStateAccessor;
+    /**
+     * The roster state singleton. This is the state that holds the candidate roster hash and the list of pairs of
+     * active roster hashes and the round number in which those rosters became active.
+     *
+     * @implNote the use of {@link ReadablePlatformStateStore} and {@link WritablePlatformStateStore} to provide access
+     * to the roster states (beyond just the {@link PlatformState}) is deliberate, for convenience.
+     */
+    private final WritableSingletonState<RosterState> rosterState;
+
+    private final WritableKVState<ProtoBytes, Roster> rosterMap;
+
+    /**
+     * Constructs a new {@link WritableRosterStore} instance.
+     *
+     * @param writableStates the readable states
+     */
+    public WritableRosterStore(@NonNull final WritableStates writableStates) {
+        this.writableStates = writableStates;
+        this.rosterStateAccessor = new ReadableRosterStore(writableStates);
+        this.rosterState = writableStates.getSingleton(RosterStateId.ROSTER_STATES_KEY);
+        this.rosterMap = writableStates.get(RosterStateId.ROSTER_KEY);
+    }
+
+    /**
+     * Set the candidate roster if valid and doesn't yet exist in the state.
+     *
+     * @param candidateRoster a candidate roster to set.
+     */
+    @Override
+    public void setCandidateRoster(@NonNull final Roster candidateRoster) {
+        Objects.requireNonNull(candidateRoster);
+        RosterValidator.validate(candidateRoster);
+
+        // update the roster state
+        final RosterState currentRosterState = rosterStateOrThrow();
+        final Bytes candidateRosterHash = RosterUtils.hash(candidateRoster).getBytes();
+        final Builder rosterStateBuilder = RosterState.newBuilder()
+                .candidateRosterHash(candidateRosterHash)
+                .roundRosterPairs(currentRosterState.roundRosterPairs());
+        this.rosterState.put(rosterStateBuilder.build());
+
+        // update the roster map and commit the changes
+        this.rosterMap.put(ProtoBytes.newBuilder().value(candidateRosterHash).build(), candidateRoster);
```

Comment:
I think there is a missing bit of logic here.

If a new candidate roster is set, but the current candidate roster isn't null, then you need to safely remove the old candidate roster from the map.